rustls/src/client/tls13.rs

```diff
@@ -997,15 +995,9 @@ impl ExpectTraffic {
         // Mustn't be interleaved with other handshake messages.
         common.check_aligned_handshake()?;
 
-        match kur {
-            KeyUpdateRequest::UpdateNotRequested => {}
-            KeyUpdateRequest::UpdateRequested => {
-                self.want_write_key_update = true;
-            }
-            _ => {
-                common.send_fatal_alert(AlertDescription::IllegalParameter);
-                return Err(Error::CorruptMessagePayload(ContentType::Handshake));
-            }
+        if common.should_update_key(key_update_request)? {
```

Comment:
Could we turn `common.should_update_key()` into `key_schedule.maybe_update_key()`, so this interface becomes simpler?

Reply:
I'm genuinely not sure if that's better, but it's slightly fewer lines and less duplication. It feels like there's something larger that isn't modeled quite right here, in that the _combination_ of common and key_schedule are needed for a bunch of behavior, and putting the logic on either of them ends up reaching into the other in a way that doesn't _feel_ quite right.

Reply:
I think I understand what you're getting at... At the same time, the `key_schedule` in this case is holding the valuable cryptographic secrets, so IMO giving it control over other parts (in particular, the `CommonState`, which is like a fairly promiscuous part of the connection that is accessed from all over the place) doesn't feel too crazy to me. (On the other hand, making another thing that gets access to the `key_schedule` feels relatively scary.)

Reply:
I'm curious what you think about the version I'm about to push. After a few shufflings in response to this feedback, I ended up moving it onto CommonState and feel relatively good about that.

Reply:
changes in 975a8f8c

Reply:
Yeah, sorry -- not such a fan of giving `CommonState` `&mut` access to the `key_schedule`. I think the design is set up to keep the `key_schedule` private within a mostly opaque state for a reason.

Reply:
Will happily invert the caller here. Is the underlying concern that malicious developers may make later changes misusing that `pub(crate)` `&mut` access? Or are there memory-safety concerns here that I should read up on?

Reply:
Any caller of this code necessarily must have `&mut` for both CommonState and KeyScheduleTraffic, so it doesn't seem like it makes any safety difference where the associated function is defined?

Reply:
No, there are no memory-safety concerns, and I think all of this is in the end maybe more a matter of taste than strict rules. However I think one of the design guidelines in *ring* and rustls is to minimize how much code gets access to secret values, and so I feel like this is in line with that idea.

(We could keep this open for a bit to see if @ctz has an opinion.)

Reply:
On 975a8f8c3 I have some probably-unhelpful thoughts:

- `KeySchedule` is being passed `CommonState` so (mainly) it can manipulate the record layer. I don't think it should use its `CommonState` reference for sending messages or much else really. Ideally I'd hope that tls13/key_schedule.rs is limited to the computations in https://www.rfc-editor.org/rfc/rfc8446#section-7.1
- When we're hoisting up identical code from client/server files, I think ideally that should go in `CommonState` as a first step. But I think the readability/future-defect improvement that can be realised from this means deduplicating tens of lines of code. There's not likely to be a code size improvement because I think it's unlikely that consumers who care about code size compile in both the client and server code into the same application.
- It's not ideal to expose `KeySchedule*` outside the state machine objects that own it, but I think in this case we could expose a closure on `key_schedule.update_encrypter`?

Reply:
It's a little hard to get from those thoughts to a concrete design. I would propose we just go back to keeping `should_update_key()` and `update_encrypter_and_notify()` separate, and we can also move the call to `common.enqueue_key_update_notification()` back into the state machine states.

In this case I agree the readability/future-defect improvement isn't huge, although I think there are benefits downstream of deduplicating/sharing logic in a single place that you might be underestimating here (I already didn't think code size was a factor here).

I think a closure would be a little to express in the type system and I generally find that the indirection caused by closures usually makes code harder to understand.